logging/exp/logger.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package exp is an experimental version of logging using slog.
+package exp
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/slog"
+)
+
+// contextKey is a private string type to prevent collisions in the context map.
+type contextKey string
+
+// loggerKey points to the value in the context where the logger is stored.
+const loggerKey = contextKey("logger")
+
+var (
+	// defaultLogger is the default logger. It is initialized once per package
+	// include upon calling DefaultLogger.
+	defaultLogger     *slog.Logger
+	defaultLoggerOnce sync.Once
+)
+
+// NewFromEnv creates a new logger from env vars.
+// Set envPrefix+"LOG_LEVEL" to overwrite log level. Default log level is warning.
+// Set envPrefix+"LOG_MODE" to overwrite log mode. Default log mode is production.
+func NewFromEnv(envPrefix string) *slog.Logger {
+	logMode := strings.ToLower(strings.TrimSpace(os.Getenv(envPrefix + "LOG_MODE")))
+	devMode := logMode == "development"
+
+	level, err := LookupLevel(os.Getenv(envPrefix + "LOG_LEVEL"))
+	if err != nil {
+		mini := slog.New(slog.NewJSONHandler(os.Stderr))
+		mini.Warn("invalid log level, defaulting to info", "error", err)
+		level = LevelInfo
```

Comment:
I don't love this. Originally I had a panic, but that felt bad too.

Reply:
Agreed, don't panic when logging. Another option is round or truncate to nearest level.